---
title: "Circulating SARS-CoV-2 RBD variants"
date: "8/5/2020"
output:
  github_document:
    toc: true
    html_preview: false
editor_options: 
  chunk_output_type: inline

---
  
This notebook analyzes RBD variants that have been sampled in isolates within the current SARS-CoV-2 pandemic. It outputs a table of mutants with the number of times each has been sampled among GISAID sequences, which can be loaded into other analyses.

## Setup

```{r setup, message=FALSE, warning=FALSE, error=FALSE}
#list of packages to install/load
packages = c("yaml","data.table","tidyverse","gridExtra","bio3d","seqinr","knitr")
#install any packages not already installed
installed_packages <- packages %in% rownames(installed.packages())
if(any(installed_packages == F)){
  install.packages(packages[!installed_packages])
}
#load packages
lapply(packages, library, character.only=T)

knitr::opts_chunk$set(echo = T)
knitr::opts_chunk$set(dev.args = list(png = list(type = "cairo")))

#read in config file
config <- read_yaml("config.yaml")

#read in file giving concordance between RBD numbering and SARS-CoV-2 Spike numbering
RBD_sites <- read.csv(config$RBD_sites,stringsAsFactors = F)

#make output directory
if(!file.exists(config$circulating_variants_dir)){
  dir.create(file.path(config$circulating_variants_dir))
}
```
Session info for reproducing environment:
```{r print_sessionInfo}
sessionInfo()
```

Read in table of mutation effects on expression and ACE2 binding

```{r read_data}

mutants <- data.table(read.csv(file=config$mut_bind_expr,stringsAsFactors = F))

# #rename mutants site indices to prevent shared names with RBD_sites, simplifying some downstream calculations that cross-index these tables
# setnames(mutants, "site_RBD", "RBD_site");setnames(mutants, "site_SARS2", "SARS2_site")

```

## Analyzing amino acid diversity in GISAID Spike sequences

We constructed an alignment of all Spike sequences available on GISAID as of 6 September, 2020. On the EpiCoV page, under downloads, one of the pre-made options is a fasta of all Spike sequences isolated thus far, which is updated each day. I have downloaded this file, unzipped, replaced spaces in fasta headers with underscores, and aligned sequences. We load in this alignment using the `read.fasta` function of the `bio3d` package, and trim the alignment to RBD residues. We remove sequecnes from non-human isolates (e.g. bat, pangolin, "environment", mink, etc.) and sequences with gap `-` characters, and then iterate through the alignment and save any observed mutations. We then filter mutations based on rationale below, and add counts of filtered observations for each mutation as an 'nobs' colum in our overall mutants data table.

We filter out any mutations that were *only* observed on sequences with large numbers of missing `X` characters -- from my initial pass, I saw some singleton amino acid variants which would require >1 nt change, and these were only found in a single sequence with many X amino acid characters (the first half of the sequence was well determined, but the second half was all X's, with the annotated "differences" being within short stretches between Xs with determiined amino acids), which made me realize I needed to be careful not only of sequences rich in gap "-" characters, but also ambiguous "X" characters. However, I didn't want to remove all sequences with undetermined characters off the bat, because another pattern I saw is that for isolates bearing the N439K mutation, >10 are well determined across the entire RBD, but ~80 have many X characters (in part of the RBD that is *not* near the N439K sequence call). So, my preference would be to believe a mutation observed in an X-rich sequence *if the variant in the sequence is observed in at least one variant that does not contain an X*, but not believe mutations that are *only* observed in X-rich sequences. (I noticed this issue with N439K, but this is not the only mutation like this which is observed on 0X sequences at least once but other times on sequences with X characters.) That is the filtering I therefore do below.


```{r gisaid_spike_alignment}
alignment <- bio3d::read.fasta(file=config$GISAID_alignment, rm.dup=T)

#remove non-human samples
keep <- grep("Human",alignment$id);  alignment$ali <- alignment$ali[keep,]; alignment$id <- alignment$id[keep]

#remove columns that are gaps in first reference sequence
alignment$ali <- alignment$ali[,alignment$ali[1,]!="-"]

alignment_RBD <- alignment; alignment_RBD$ali <- alignment$ali[,RBD_sites$site_SARS2]

#check that the first sequence entry matches our reference RBD sequence
stopifnot(sum(!(alignment_RBD$ali[1,] == RBD_sites[,"amino_acid_SARS2"]))==0)

#remove sequences have gaps, as the amino acid calls may be generally unreliable
remove <- c()
for(i in 1:nrow(alignment_RBD$ali)){
  if(sum(alignment_RBD$ali[i,]=="-") > 0){remove <- c(remove,i)}
}

alignment_RBD$ali <- alignment_RBD$ali[-remove,];alignment_RBD$id <- alignment_RBD$id[-remove]

#output all mutation differences from the WT/reference RBD sequence
#I do this by iterating over rows and columns of the alignment matrix which is STUPID but effective
variants_vec <- c()
isolates_vec <- c()
for(j in 1:ncol(alignment_RBD$ali)){
  #print(i)
  for(i in 1:nrow(alignment_RBD$ali)){
    if(alignment_RBD$ali[i,j] != alignment_RBD$ali[1,j] & !(alignment_RBD$ali[i,j] %in% c("X","-"))){
      variants_vec <- c(variants_vec, paste(alignment_RBD$ali[1,j],j,alignment_RBD$ali[i,j],sep=""))
      isolates_vec <- c(isolates_vec, alignment_RBD$id[i])
    }
  }
}

#remove any mutations that are *only* observed in X-rich sequences of dubious quality (keep counts in X-rich sequences if they were observed in at least one higher quality isolate)
#make a data frame that gives each observed mutation, the isolate it was observed in, and the number of X characters in that sequence. Also, parse the header to give the country/geographic division of the sample
variants <- data.frame(isolate=isolates_vec,mutation=variants_vec)
for(i in 1:nrow(variants)){
  variants$number_X[i] <- sum(alignment_RBD$ali[which(alignment_RBD$id == variants[i,"isolate"]),]=="X")
  variants$geography[i] <- strsplit(as.character(variants$isolate[i]),split="/")[[1]][2]
}
#filter the sequence set for mutations observed in at least one X=0 background
variants_filtered <- data.frame(mutation=unique(variants[variants$number_X==0,"mutation"])) #only keep variants observed in at least one sequence with 0 X
for(i in 1:nrow(variants_filtered)){
  variants_filtered$n_obs[i] <- sum(variants$mutation == variants_filtered$mutation[i]) #but keep counts for any sequence with observed filtered muts
  variants_filtered$n_geography[i] <- length(unique(variants[variants$mutation == variants_filtered$mutation[i],"geography"]))
  variants_filtered$list_geography[i] <- list(list(unique(variants[variants$mutation == variants_filtered$mutation[i],"geography"])))
}

#add count to mutants df
mutants[,nobs:=0]
mutants[,ngeo:=0]
mutants[,geo_list:=as.list(NA)]
for(i in 1:nrow(mutants)){
  if(mutants$mutation_RBD[i] %in% variants_filtered$mutation){
    mutants$nobs[i] <- variants_filtered[variants_filtered$mutation==mutants$mutation_RBD[i],"n_obs"]
    mutants$ngeo[i] <- variants_filtered[variants_filtered$mutation==mutants$mutation_RBD[i],"n_geography"]
    mutants$geo_list[i] <- variants_filtered[variants_filtered$mutation==mutants$mutation_RBD[i],"list_geography"]
  }
}


```

We see `r sum(mutants$nobs)` amino acid polymorphisims within the `r nrow(alignment_RBD$ali)` sequences uploaded in GISAID, which represents `r sum(mutants$nobs>0)` of our `r nrow(mutants[mutant!=wildtype & mutant!="*",])` measured missense mutants. In the table below, we can see that many of these mutations are observed only one or a few times, so there may still be unaccounted for sequencinig artifacts, which we tried to account for at least minimally with some filtering above.

```{r table_circulating_variants_nobs}
kable(table(mutants[mutant!=wildtype & mutant!="*",nobs]),col.names=c("mutation count","frequency"))
```

For curiosity's sake, here are tables giving mutations that were seen >20 times, and those seen any number of times with measured binding effects >0.1.

```{r table_most_common_variants, echo=F}
kable(mutants[nobs>20,.(mutation,expr_lib1,expr_lib2,expr_avg,bind_lib1,bind_lib2,bind_avg,nobs,ngeo)],
      col.names=c("Mutation","expr, lib1","expr, lib2","expression effect","bind, lib1","bind, lib2", "binding effect","number of GISAID sequences", "number locations"))
```

```{r table_highest_binding_variants, echo=F}
kable(mutants[nobs>0 & bind_avg>0.1,.(mutation,expr_lib1,expr_lib2,expr_avg,bind_lib1,bind_lib2,bind_avg,nobs,ngeo)],
      col.names=c("Mutation","expr, lib1","expr, lib2","expression effect","bind, lib1","bind, lib2", "binding effect","number of GISAID sequences", "number locations"))
```

What are the sequences on which these stronger affinity-enhancing muts have been sampled?

```{r affinity-enhancing}
mutants[bind_avg>0.2 & nobs>0,]

alignment_RBD$id[alignment_RBD$ali[,171]=="Y"]
alignment_RBD$id[alignment_RBD$ali[,123]=="F"]
```

Illustrate table showing any circulating mutations found at any our sites of significant antigenic escape across all mAbs.

```{r antigenic-positions}
sig_sites <- read.csv(config$significant_escape_sites)

kable(mutants[site_SARS2 %in% sig_sites$site & nobs>0,c("mutation","bind_avg","expr_avg","nobs","ngeo","geo_list")],
      col.names=c("Mutation","binding effect","expression effect","number of GISAID sequences", "number locations","locations"))
```

Output table giving nobs and geographic spread for circulating mutants.

```{r output_GISAID_counts}
write.csv(mutants[,.(site_RBD,site_SARS2,wildtype,mutant,mutation,nobs,ngeo)],file=config$circulating_variants,row.names=F)

```
